File: internal/game/states/town.go
package states

import (
	"github.com/slycrel/slycrel/internal/game"
	"github.com/slycrel/slycrel/internal/model"
)

// TownState shows the town ANSI menu.
// Original: ST_Town
type TownState struct{}

func (TownState) ID() game.StateID { return "town" }

func (TownState) Enter(s *game.Session) {
	if s.Character != nil {
		s.Character.Location = model.TheTown
	}
	s.IO.ShowANSIFile("main_menu")
	s.SetNextState("town_prompt")
}

// TownPromptState shows the town menu prompt and routes to locations.
// Original: ST_TownShortPrompt + ST_GetTownMenu
type TownPromptState struct{}

func (TownPromptState) ID() game.StateID { return "town_prompt" }

func (TownPromptState) Enter(s *game.Session) {
	s.IO.Outln("[A,B,C,H,I,J,L,S,T,V,W,X,@,?]", true, 1)
	s.IO.Cr()

	choice := s.IO.LettersPrompt("Your Choice?", "ABCHIJLSTVWX@?", 1, true, true)

	switch choice {
	case "A":
		s.Character.Location = model.TheArenaMenu
		s.SetNextState("arena")
	case "B":
		s.Character.Location = model.TheBank
		s.SetNextState("bank")
	case "C":
		s.Character.Location = model.TheCommonGuild
		s.SetNextState("common_guild")
	case "H":
		s.Character.Location = model.TheHerbalist
		s.SetNextState("herbalist")
	case "I":
		inn, _ := s.Store.LoadInn()
		if inn != nil && !inn.Open && inn.Owner != s.Character.Name {
			s.IO.Outln("The Inn's windows are boarded up, with a sign painted on them--KEEP OUT!", true, 3)
			s.IO.Cr()
			s.IO.Outln("(This means that it's closed...)", true, 1)
			s.SetNextState("town_prompt")
		} else {
			s.Character.Location = model.TheInn
			s.SetNextState("inn")
		}
	case "J":
		s.Character.Location = model.TheJail
		s.SetNextState("jail")
	case "L":
		s.Character.Location = model.TheArmory
		s.SetNextState("armory")
	case "S":
		s.Character.Location = model.TheBlacksmiths
		s.SetNextState("blacksmith")
	case "T":
		s.Character.Location = model.TheTavern
		s.SetNextState("tavern")
	case "V":
		s.SetReturnState("town_prompt")
		s.SetNextState("view_character")
	case "W":
		s.Character.Location = model.TheWildernessMenu
		s.SetNextState("wilderness")
	case "X":
		s.SetNextState("quit")
	case "@":
		s.Character.Location = model.TheTower
		s.SetNextState("tower")
	case "?":
		s.SetNextState("town")
	default:
		s.SetNextState("town_prompt")
	}
}

File: clients/web/game/states/town.js
import { WhereType } from '../../model/enums.js';
import { saveCharacter } from '../../store/local.js';
import { BeginState } from './begin.js';
import { BankState } from './bank.js';

// Mapping of the town menu's letter keys to human-readable location names,
// used by the stub message until each location is ported.
const LOCATION_BY_KEY = {
  A: { name: 'Arena',          where: WhereType.TheArenaMenu },
  B: { name: 'Bank',           where: WhereType.TheBank },
  C: { name: 'Common Guild',   where: WhereType.TheCommonGuild },
  H: { name: 'Herbalist',      where: WhereType.TheHerbalist },
  I: { name: 'Inn',            where: WhereType.TheInn },
  J: { name: 'Jail',           where: WhereType.TheJail },
  L: { name: 'Armory',         where: WhereType.TheArmory },
  S: { name: 'Blacksmith',     where: WhereType.TheBlacksmiths },
  T: { name: 'Tavern',         where: WhereType.TheTavern },
  W: { name: 'Wilderness',     where: WhereType.TheWildernessMenu },
  '@': { name: 'Tower',        where: WhereType.TheTower },
};

// TownState — renders main_menu.ans (the town map ANSI screen).
// Mirrors internal/game/states/town.go TownState.
export class TownState {
  async enter(session) {
    if (session.character) session.character.location = WhereType.TheTown;
    session.io.clear();
    await session.io.showAnsiFile('main_menu');
    session.setNext(new TownPromptState());
  }
}

// TownPromptState — reads the menu choice and routes (or stubs).
// Mirrors internal/game/states/town.go TownPromptState.
export class TownPromptState {
  async enter(session) {
    const { io } = session;
    io.println('[A,B,C,H,I,J,L,S,T,V,W,X,@,?]', 1);
    io.cr();

    const choice = await io.lettersPrompt('Your Choice?', 'ABCHIJLSTVWX@?');
    io.cr();

    if (choice === '?') {
      session.setNext(new TownState());
      return;
    }
    if (choice === 'V') {
      // View character — defer to a real port once view_character is callable
      // from the town menu. For now, stub like the rest of the locations.
      io.println('(stub) character sheet from town menu not yet implemented.', 6);
      await io.pausePrompt();
      session.setNext(new TownPromptState());
      return;
    }
    if (choice === 'X') {
      io.println('Saving and returning to the gateway...', 5);
      saveCharacter(session.character);
      await io.pausePrompt();
      session.setNext(new BeginState());
      return;
    }

    if (choice === 'B') {
      session.setNext(new BankState());
      return;
    }

    const loc = LOCATION_BY_KEY[choice];
    if (loc) {
      session.character.location = loc.where;
      io.println(`(stub) ${loc.name} not yet implemented.`, 6);
      await io.pausePrompt();
      session.setNext(new TownPromptState());
      return;
    }

    session.setNext(new TownPromptState());
  }
}
